### PR TITLE
Move generation isoscape functions to new file; implement dispatch_rasters

### PR DESCRIPTION
**Pull request by VimWizardVersace**
To prepare for generating rasters using EE, refactor file structure and move isoscape generation to dedicated file. I noticed eeraster.py depends on raster.py, so housing isoscape generation in raster.py creates a circular dependency.

Also implement a new function, dispatch_rasters, to decide whether to use ee or local files for sourcing rasters.

### CODE REVIEW

File: generate_isoscape.py
Comment (rothn) [inline generate_isoscape.py:187]: Using the `np` name for a local variable violates the widely-held convention of reserving `np` for the numpy import. Please change the name of this variable.

**Review by rothn — CHANGES_REQUESTED** [review]
In general, this looks good. I have several requests for clarification and a few change requests before we merge.

Please remember to stage any changes in other repos and merge at the same time. Please also be ready to revert immediately if this causes any breakage (reverting the revert should be _very_ easy).

Since this is a ddfcommon change at a critical time, @benwulfe should also sign off on this once the outstanding comments are resolved.

File: generate_isoscape.ipynb
Comment (VimWizardVersace) [reply generate_isoscape.ipynb:1]: It's bundled with the saved model, and loaded by model.Model()

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:79]: I agree. I resolve this in a follow up PR.

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:58]: Great question! Something from old versions of the code I never bothered to question until now. 

Fixed

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:49]: raster_test.ipynb has some tests that load a bundled TFModel and ColumnTransformers to build isoscapes. I think beyond initial development (and after adding tests), they've sparked joy and been easy to reason about.

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:22]: We probably load rasters of all different projections. I'm not opinionated on how we project our generated rasters though, as we load rasters from earth engine which we can reproject how we please.

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:64]: I'm having a hard time figuring out how to meaningfully validate this. The column order is enforced by the order of the `geotiffs` dictionary. I feel like checking if it matches the order in the model's ColumnTransformer is redundant as `geotiffs` is created by iterating over all the columns in ColumnTransformer + the dictionary's order is preserved.

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:119]: Unfortunately this would break the pipeline. We should first reproject all our dependencies before making a change like that.

File: generate_isoscape.py
Comment (rothn) [reply generate_isoscape.py:22]: WGS84 is a good starting place-- I just want to make sure we are consistent in whatever we choose. We may want to ask Martinelli his opinion at some point to minimize distortion.

File: generate_isoscape.ipynb
Comment (rothn) [reply generate_isoscape.ipynb:1]: That looks good.

File: generate_isoscape.py
Comment (rothn) [reply generate_isoscape.py:64]: Since this is an expensive function, it seems like it should be possible to check that the order of `geotiffs.keys()` matches the order of the feature keys listed by the model. This stuff tends to break over time, and having a check at the top of the function would preempt that (or even better-- extracting the check into a function and using it to guard API calls).

File: generate_isoscape.py
Comment (rothn) [reply generate_isoscape.py:117]: Could we at least print a warning and announce that this will be an exception in the future?

File: eeraster.py
Comment (rothn) [reply eeraster.py:286]: SGTM

File: generate_isoscape.py
Comment (rothn) [reply generate_isoscape.py]: I don't see the change-- did you remember to push?

**Review by rothn — CHANGES_REQUESTED** [review]
Just a few follow-ups. Thanks again!

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py]: It's there

`if band.CreateMaskBand(0) > gdal.CE_Warning:`

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:71]: Unfortunately just ran into this:
`MaskError: Mask and data not compatible: data size is 180000, mask size is 90000.`

so i reverted it back to two masks. getting around this does not seem to be trivial nor worth it right now.

File: generate_isoscape.py
Comment (rothn) [inline generate_isoscape.py:135]: Why is this a nested method as opposed to the alternative of leaving it at module-scope? Is there anywhere else that we handle a list of rasters? If so, that may also be an opportunity to use this functionality. Otherwise, this probably makes sense. Curious what you have in mind.

**Review by rothn — CHANGES_REQUESTED** [review]
Getting close! Remember to have all changes ready/staged to commit at the same time, and get approval from @benwulfe since this is a time of transition. Please also be ready to revert immediately if this causes problems (that can easily be undone later).

**Review by benwulfe — CHANGES_REQUESTED** [review]
please also see the code at the top of eeddf.py, which reloads all modules.
Can you please add generate_isoscape.py to this list:
https://github.com/tnc-br/ddf_common/blob/main/eeddf.py#L10

and also run the ddf_ee_api.ipynb if you can to make sure it still works -- which is our public notebook that documents how to use the external api.
If you make a new public API, please expose it thru ddf_ee_api.

File: generate_isoscape.py
Comment (VimWizardVersace) [reply generate_isoscape.py:135]: I can't see it being used anywhere else, but i moved it to module-level scope to be consistent with the rest of the file.

**Comment by VimWizardVersace** [comment]
> please also see the code at the top of eeddf.py, which reloads all modules. Can you please add generate_isoscape.py to this list: https://github.com/tnc-br/ddf_common/blob/main/eeddf.py#L10
> 
> and also run the ddf_ee_api.ipynb if you can to make sure it still works -- which is our public notebook that documents how to use the external api. If you make a new public API, please expose it thru ddf_ee_api.

I ran it and it worked, but i don't think anything in ddf_ee_api.ipynb intersects with any of these changes. I test it by running generate_isoscape.ipynb

File: generate_isoscape.py
Comment (rothn) [reply generate_isoscape.py:58]: I don't see anything in raster_test.ipynb that checks for this case. Column Transformers and data preprocessing in general often cause trouble-- if we don't beef these up now, we'll probably spend a good amount of time bug-hunting later. I'm not going to block on this, but a little time now will likely save us more time down the road.

**Review by rothn — APPROVED** [review]

**Review by benwulfe — DISMISSED** [review]

File: test_ee_raster.ipynb
Comment (rothn) [inline test_ee_raster.ipynb:1]: Looks like this got corrupted by the merge. We will need to merge manually or replace this with something we know works.

**Review by rothn — CHANGES_REQUESTED** [review]
Looks like test_ee_raster.ipynb got corrupted. We should fix that before pushing to main.

**Review by rothn — APPROVED** [review]

**Pull request merged**
